modules/irqchip/riscv/fdt_irqchip.c

```diff
@@ -11,6 +11,358 @@
+
+   val = fdt_getprop(fdt, node, "interrupts", &len);
+   if (!val)
+      return -ENODEV;
+
+   // Check if the index is within the parameter range
+   if (index * ints_cells >= len / sizeof(*val))
+      return -EINVAL;
+
+   for (u32 i = 0; i < ints_cells; i++)
+      irq_param->params[i] = fdt32_to_cpu(val[index * ints_cells + i]);
+
+   irq_param->intc_node = intc_node;
+   irq_param->param_nums = ints_cells;
+
+   return 0;
+}
+
+int irqchip_get_irq_count(void *fdt, int node)
+{
+   struct fdt_irq_param irq_param;
+   int index = 0;
+   int rc = 0;
+
+   while (rc == 0) {
+      rc = fdt_parse_one_hwirq(fdt, node, index, &irq_param);
+      index++;
+   }
+
+   return index;
+}
+
+int irqchip_alloc_irq(void *fdt, int node, int hwirq_index)
+{
+   struct irq_domain *domain, *tmp;
+   struct fdt_irq_param irq_param;
+   int irq = 0, rc;
+   int hwirq, type;
+
+   rc = fdt_parse_one_hwirq(fdt, node, hwirq_index, &irq_param);
+   if (rc)
+      return 0;
+
+   disable_preemption();
+
+   /* Find the corresponding irq domain */
+   list_for_each(domain, tmp, &irq_domains, node) {
+
+      if (domain->fdt_node == irq_param.intc_node) {
+
+         if (domain->ops->xlate_irq_param) {
+            rc = domain->ops->xlate_irq_param(&irq_param, &hwirq,
+                                              &type, domain->priv);
+         } else {
+            rc = default_xlate_irq_param(&irq_param, &hwirq,
+                                         &type, NULL);
+         }
+
+         if (rc) {
+            enable_preemption();
+            return 0;
+         }
+
+         irq = irqchip_get_free_irq(domain, hwirq);
+         if (irq < 0) {
+            enable_preemption();
+            return -EINVAL;
+         }
+
+         if (domain->ops->hwirq_set_type)
+            domain->ops->hwirq_set_type(hwirq, type, domain->priv);
+
+         domain->irq_map[hwirq] = irq;
+         break;
+      }
+   }
+
+   enable_preemption();
+   return irq;
+}
+
+void irqchip_free_irq(int irq)
+{
+   struct irq_domain *domain, *tmp;
+
+   disable_preemption();
+
+   list_for_each(domain, tmp, &irq_domains, node) {
+
+      for (u32 i = 0; i < domain->irq_map_size; i++) {
+
+         if (domain->irq_map[i] == irq) {
+
+            domain->irq_map[i] = 0;
+            irqchip_put_irq(irq);
+            enable_preemption();
+            return;
+         }
+      }
+   }
+
+   enable_preemption();
+}
+
+struct irq_domain *
+irqchip_register_irq_domain(void *priv,
+                            int fdt_node,
+                            int int_nums,
+                            struct fdt_irqchip_ops *ops)
+{
+   struct irq_domain *domain;
+
+   domain = kzmalloc(sizeof(*domain) + int_nums * sizeof(int));
+   if (!domain)
+      return NULL;
+
+   domain->fdt_node = fdt_node;
+   domain->irq_map_size = int_nums;
+   domain->ops = ops;
+   domain->priv = priv;
+
+   disable_preemption();
+   list_add_tail(&irq_domains, &domain->node);
+   enable_preemption();
+
```

Comment:
Extra empty line before the `return` statement.

Reply:
fixed！